write_summary/src/progress.rs

```diff
@@ -120,6 +141,32 @@ impl SequencerProgress {
     }
 }
 
+// Ensures that `buffered` is less than `actively_buffering`,
+// returning the new buffered value
+fn clamp(
+    buffered: SequenceNumber,
+    actively_buffering: Option<SequenceNumber>,
+) -> Option<SequenceNumber> {
+    let completed_buffering = if let Some(val) = actively_buffering {
+        let val = val.get();
+        if val > 0 {
+            SequenceNumber::new(val - 1)
+        } else {
+            // no sequece number has completed buffering yet
+            return None;
+        }
```

Comment:
Nitpick: instead of checking for a potential underflow and then performing the "dangerous" action, you can just do that in one go (which is also safer if someone refactors the code):

```suggestion
        // returns `None` if no sequence number has completed buffering yet
        SequenceNumber::new(val.get().checked_sub(1)?)
```

Note here that `?` also works for `Option`s.

Reply:
Nice! TIL `checked_sub` !

Reply:
Tests for the win -- turns out `SequenceNumber` is actually backed by a signed integer so `checked_sub` doesn't underflow:


```
---- progress::tests::test_clamp stdout ----
thread 'progress::tests::test_clamp' panicked at 'assertion failed: `(left == right)`
  left: `Some(SequenceNumber(-1))`,
 right: `None`', write_summary/src/progress.rs:394:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::assert_failed_inner
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:225:17
   3: core::panicking::assert_failed
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:182:5
   4: write_summary::progress::tests::test_clamp
             at ./src/progress.rs:394:9
   5: write_summary::progress::tests::test_clamp::{{closure}}
             at ./src/progress.rs:386:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```


I tweaked it in 3f2084bcc